gateway-server/src/main/java/org/apache/knox/gateway/services/token/impl/state/MultiFileTokenStateJournal.java

```diff
@@ -0,0 +1,142 @@
+import java.io.OutputStream;
+import java.io.OutputStreamWriter;
+import java.nio.channels.Channels;
+import java.nio.channels.FileChannel;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A TokenStateJournal implementation that manages separate files for token state.
+ */
+class MultiFileTokenStateJournal extends FileTokenStateJournal {
+
+    // File extension for journal entry files
+    static final String ENTRY_FILE_EXT = ".ts";
+
+    // Filter used when listing all journal entry files in the journal directory
+    static final String ENTRY_FILE_EXT_FILTER = "*" + ENTRY_FILE_EXT;
+
+    MultiFileTokenStateJournal(GatewayConfig config) throws IOException {
+        super(config);
+    }
+
+    @Override
+    public void add(final String tokenId, long issueTime, long expiration, long maxLifetime) throws IOException {
+        add(Collections.singletonList(new FileJournalEntry(tokenId, issueTime, expiration, maxLifetime)));
+    }
+
+    @Override
+    public void add(final List<JournalEntry> entries) throws IOException {
+        // Persist each journal entry as an individual file in the journal directory
+        for (JournalEntry entry : entries) {
+            final Path entryFile = journalDir.resolve(entry.getTokenId() + ENTRY_FILE_EXT);
+            log.persistingJournalEntry(entryFile.toString());
+            try (FileChannel fileChannel = FileChannel.open(entryFile, StandardOpenOption.WRITE,
+                    StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING)) {
+                fileChannel.lock();
+                try (OutputStream out = Channels.newOutputStream(fileChannel)) {
+                    BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(out, StandardCharsets.UTF_8));
+                    writer.write(entry.toString());
+                    writer.newLine();
+                    writer.flush();
+                }
+                log.addedJournalEntry(entry.getTokenId());
+            } catch (IOException e){
+                log.failedToPersistJournalEntry(entry.getTokenId(), e);
+                throw e;
+            }
+        }
+    }
+
+    @Override
+    public JournalEntry get(final String tokenId) throws IOException {
+        JournalEntry result = null;
+
+        Path entryFilePath = journalDir.resolve(tokenId + ENTRY_FILE_EXT);
+        if (Files.exists(entryFilePath)) {
+            try (FileChannel fileChannel = FileChannel.open(entryFilePath, StandardOpenOption.READ)) {
+                fileChannel.lock(0L, Long.MAX_VALUE, true);
+                List<FileJournalEntry> entries = loadJournal(fileChannel);
+                if (entries.isEmpty()) {
+                    log.journalEntryNotFound(tokenId);
```

Comment:
This is a different case than what we have in line 99 (the file is there but empty). I'd create a different log entry for this case (to help us when debugging).